From: Duncan Eastoe <duncan.eastoe@att.com>
Date: Mon, 28 Sep 2020 12:51:19 +0100
Subject: zebra: fix FPM abort on blackhole/unreach/prohibit

`netlink_route_info_fill()` returns 0 (indicating failure) for
any type of blackhole route. This causes an early return from
`zfpm_netlink_encode_route()` and thus ultimately triggers an
assertion failure in `zfpm_build_route_updates()`.

This is partially fixed (for 7.5) in upstream commit
b0e9567ed162da708f8d0b3a3caf87cd03b62e96. However with this
fix in place the issue is still observed with unreachable and
prohibit routes.

To completely address the issue `netlink_route_info_fill()` is
updated to not indicate failure, due to lack of nexthops, for
any blackhole routes.

Signed-off-by: Duncan Eastoe <duncan.eastoe@att.com>
---
 zebra/zebra_fpm_netlink.c | 14 +++++++++++---
 1 file changed, 11 insertions(+), 3 deletions(-)

diff --git a/zebra/zebra_fpm_netlink.c b/zebra/zebra_fpm_netlink.c
index 9253ef0..f47ac5e 100644
--- a/zebra/zebra_fpm_netlink.c
+++ b/zebra/zebra_fpm_netlink.c
@@ -372,10 +372,18 @@ static int netlink_route_info_fill(netlink_route_info_t *ri, int cmd,
 		}
 	}
 
-	/* If there is no useful nexthop then return. */
 	if (ri->num_nhs == 0) {
-		zfpm_debug("netlink_encode_route(): No useful nexthop.");
-		return 0;
+		switch (ri->rtm_type) {
+		case RTN_PROHIBIT:
+		case RTN_UNREACHABLE:
+		case RTN_BLACKHOLE:
+			break;
+		default:
+			/* If there is no useful nexthop then return. */
+			zfpm_debug(
+				"netlink_encode_route(): No useful nexthop.");
+			return 0;
+		}
 	}
 
 	return 1;
